카카오로그인 기능 구현

Nitoriubar opened the pull request:
## 📝 세부사항
- Kakao Login 
- 회원가입 & 로그인 구현

## 📗 참고사항


## 📎 관련 이슈


close #5

Nitoriubar committed:
message -add login API

Nitoriubar committed:
Create KakaoLogin API

hossi97 commented on src/main/java/com/triprint/backend/domain/user/config/SecurityConfig.java:
```diff
@@ -0,0 +1,66 @@
+@Configuration
+@EnableWebSecurity
+@RequiredArgsConstructor
+public class SecurityConfig extends WebSecurityConfigurerAdapter {
+
+    @Autowired
+    UserRepository userRepository;
+
+    public static final String FRONT_URL = "http://localhost:3000";
+
+    private final CorsFilter corsFilter;
+
+    // @Bean -> 해당 메소드의 리턴되는 오브젝트를 IoC로 등록해줌
+    @Bean
+    public BCryptPasswordEncoder encodePwd() {
+        return new BCryptPasswordEncoder();
+    }
+
+    @Override
+    protected void configure(HttpSecurity http) throws Exception {
+        http.csrf().disable()
+                .sessionManagement()  // session 을 사용하지 않음
```
WebSecurityConfigiurerAdapter 를 상속받는 방식은 Security 5.7.0 M2 버전부터 deprecated 돼서 다른 방식으로 Security Config 를 사용하는 걸로 알고 있습니다.

시간되시면 아래 글 참고하시면 좋을 것 같아요!

https://spring.io/blog/2022/02/21/spring-security-without-the-websecurityconfigureradapter

Nitoriubar replied:
감사합니다! 읽고 정리하겠습니다.

hossi97 commented on src/main/java/com/triprint/backend/domain/user/controller/UserController.java:
```diff
@@ -0,0 +1,47 @@
+
+import javax.servlet.http.HttpServletRequest;
+
+@RestController
+@RequestMapping("api")
+public class UserController {
+
+    @Autowired
+    private UserService userService;
+
```
의존관계 주입 방식에는

- 생성자 주입 방식
- 수정자 주입 방식
- 필드 주입 방식

3가지가 있습니다. 

선희님께서 사용하신 방법은 필드 주입 방식인데, 몇가지 이유로 수정자 주입이나 필드 주입보다는 생성자 주입 방식을 주로 사용합니다.

관련 내용 찾아보시면 좋을것 같습니다.

Nitoriubar replied:
넵, 알겠습니다.
확인하겠습니다.

hossi97 commented on src/main/java/com/triprint/backend/domain/user/entity/kakaoLoginDto/KakaoProfile.java:
```diff
@@ -0,0 +1,40 @@
+
+import lombok.Data;
+
+@Data
+public class KakaoProfile {
+
+    public Long id;
+    public String connected_at;
+    public Properties properties;
+    public KakaoAccount kakao_account;
+
```
@Data 어노테이션은 Getter, Setter, EqualAndHashcode 등의 여러 어노테이션의 집합인데요. 

@Setter 는 외부에서 객체의 멤버변수 값을 변경할 수 있기 때문에 잘 사용하지 않고, @EqualAndHashcode 가 예상과 달리 동작해서 에러를 야기할 수 있다고 알고 있습니다.

@Data 가 꼭 필요한게 아니라면,
필요한 어노테이션만 별도로 만들어주는 것이 좋을 것 같습니다.

그리고 개인적으로 궁금한 것이 있는데 class 내부에 class 를 정의한 이유가 있을까여?

hossi97 commented on src/main/java/com/triprint/backend/domain/user/service/UserService.java:
```diff
@@ -0,0 +1,160 @@
+
+        // JSON 응답을 객체로 변환
+        ObjectMapper objectMapper = new ObjectMapper();
+        KakaoProfile kakaoProfile = null;
+        try {
+            kakaoProfile = objectMapper.readValue(kakaoProfileResponse.getBody(), KakaoProfile.class);
+        } catch (JsonProcessingException e) {
+            e.printStackTrace();
+        }
+
+        return kakaoProfile;
```
특별한 이유가 없다면, 그리고 에러를 어차피 잡아주신다면 KakaoProfile 객체를 try문 내부에 선언과 정의를 모두 해주시는게 좀 더 코드가 간결해질 것 같습니다

hossi97 commented on src/main/java/com/triprint/backend/domain/user/service/UserService.java:
```diff
@@ -0,0 +1,160 @@
+    public User getUser(HttpServletRequest request) {
+        String userCode = (String) request.getAttribute("userCode");
+
+        User user = userRepository.findBykakaoId(userCode);
+
+        return user;
+    }
```
요런 return 문도 user 변수를 따로 return 하지 않고,
userRepository.findByKakaoId() 를 바로 return 해주시면 코드가 더 간결해질 것 같습니다.

Nitoriubar replied:
코드가 간결해질 순 있지만 추후 수정에는 변수를 리턴하는 편이 더 나을 것이라고 생각됩니다.

hossi97 commented on src/main/java/com/triprint/backend/domain/user/service/UserService.java:
```diff
@@ -0,0 +1,160 @@
+    @Autowired
+    UserRepository userRepository;
+
+    //환경 변수 가져오기
+    @Value("355bc84f673d4cd4b5d9c1b88838cf32")
+    String client_id;
+
```
yml 파일에 값을 넣으셨으니까 여기에 직접 값을 넣기보다 환경변수 ${CLIENT_ID} 로 변경해주시면 좋을것 같습니다.

Nitoriubar replied:
넵, 수정하겠습니다.

hossi97 reviewed (commented):
바쁘신 와중에 고생하셨습니다 선희님.
제가 공부한 내용 위주로 리뷰남겼는데, 의문드시는 부분들이 있다면 편하게 말씀주세요.
몸조리 잘하세요!

sangrimlee commented on src/main/java/com/triprint/backend/domain/user/controller/UserController.java:
```diff
@@ -0,0 +1,47 @@
+        HttpHeaders headers = new HttpHeaders();
+        headers.add(JwtProperties.HEADER_STRING, JwtProperties.TOKEN_PREFIX + jwtToken);
+
+        return ResponseEntity.ok().headers(headers).body("success");
```
token을 Header에 담아서 보내주시는게 아니라 body에 담아서 넘겨주셔야됩니다.

Header에 담아서 보내는 건 요청하는 부분에서 하는 거라서 body로 json형태로 보내주시면 되요.

그리고 재호님이랑 이야기해서 body의 response 구조도 한 번 이야기해보시는 것 도 좋을 것 같습니다.

Nitoriubar committed:
fix: 오류 수정

- 토큰을 Header에서 Body로 넘겨주도록 수정
- DB를 H2에서 MySQL로 수정
- FRONT_URL 수정

Nitoriubar committed:
style : 환경변수 추가

- DB_URL
- DB_NAME
- DB_PW
- FRONT_URL

Nitoriubar committed:
fix: 정상적인 요청의 경우 문제 수정

- UserService에서 정상으로 토큰이 생성된 경우 user를 null로 받아오는
문제
- UserDto를 생성 후 user의 이메일, 이름, 프로필이미지를 받아오도록 수정

Nitoriubar committed:
fix: 정상적인 동작 발생하던 에러 수정

UserController에서 정상적인 동작인 경우 Body에 사용자의 이메일과 이름, 프로필이미지를
넘겨주도록 수정

Nitoriubar committed:
style: 변수명 수정

재호님과 변수명을 밪추기 위해 DB_PW -> DB_PASSWORD로 변경

Nitoriubar committed:
Merge branch 'dev' into feat/login
